File: src/791.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_VAL_SIZE (26)  // 1 <= order.length <= 26
int val[MAX_VAL_SIZE];
int compareInteger(const void* n1, const void* n2) {
    // descending order
    return (val[*(char*)n1 - 'a'] < val[*(char*)n2 - 'a']);
}
char* customSortString(char* order, char* s) {
    char* pRetVal = NULL;

    memset(val, 0, sizeof(val));

    int orderSize = strlen(order);
    int idx;
    int i;
    for (i = 0; i < orderSize; ++i) {
        idx = order[i] - 'a';
        val[idx] = MAX_VAL_SIZE - i;
    }

    int sSize = strlen(s);
    qsort(s, sSize, sizeof(char), compareInteger);
    pRetVal = s;

    return pRetVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (200)
    struct testCaseType {
        char order[26];
        char s[MAX_SIZE];
    } testCase[] = {{"cba", "abcd"}, {"bcafg", "abcd"}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input:  order = "cba", s = "abcd"
     *  Output:  "cbad"
     *
     *  Input:  order = "bcafg", s = "abcd"
     *  Output:  "bcad"
     */

    char* pAnswer = NULL;
    int i;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: order = \"%s\", s = \"%s\"\n", testCase[i].order, testCase[i].s);

        pAnswer = customSortString(testCase[i].order, testCase[i].s);
        printf("Output: \"%s\"\n", pAnswer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/791.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def customSortString(self, order: str, s: str) -> str:
        retVal = ""

        val = defaultdict(int)
        for i, ch in enumerate(order):
            val[ch] = 26 - i    # 1 <= order.length <= 26
        retVal = "".join(sorted(s, key=lambda ch: val[ch], reverse=True))

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for order, s in zip(["cba", "bcafg"], ["abcd", "abcd"]):
            # /* Example
            #  *  Input:  order = "cba", s = "abcd"
            #  *  Output:  "cbad"
            #  *
            #  *  Input:  order = "bcafg", s = "abcd"
            #  *  Output:  "bcad"
            #  */
            logging.info("Input: order = \"%s\", s = \"%s\"", order, s)

            retVal = pSolution.customSortString(order, s)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
